packages/flutter_tools/lib/src/web/devfs_config.dart

```diff
@@ -0,0 +1,261 @@
+        throw YamlException(
+          '$devConfigFilePath file found, but it must be a YAML map (e.g., "server:"). Found a ${contents.runtimeType} instead.',
+          contents.span,
+        );
+      }
+
+      if (!contents.containsKey('server') || contents['server'] is! YamlMap) {
+        // final SourceSpan span = (contents.containsKey('server') && contents['server'] is YamlNode)
+        //     ? (contents['server'] as YamlNode).span
+        //     : contents.span;
+        throw YamlException(
+          '"$devConfigFilePath" file found, but the "server" key is missing or malformed. It must be a YAML map.',
+          null,
+        );
+      }
+
+      final serverYaml = contents['server'] as YamlMap;
+      fileConfig = WebDevServerConfig.fromYaml(serverYaml);
+      globals.printStatus('\nLoaded configuration from $devConfigFilePath');
+      globals.printTrace(fileConfig.toString());
+    } on YamlException catch (e) {
+      globals.printError('Error: Failed to parse $devConfigFilePath: ${e.message} ${e.span}');
+      rethrow;
+    } on Exception catch (e) {
+      globals.printError('An unexpected error occurred while reading $devConfigFilePath: $e');
+      globals.printStatus(
+        'Reverting to default flutter_tools web server configuration due to unexpected error.',
+      );
+    }
+  }
+
+  final String finalHost =
+      _getOverriddenValue<String>('host', fileConfig.host, hostname) ?? 'localhost';
+  final int? finalPort = _getOverriddenValue<int>(
+    'port',
+    fileConfig.port,
+    port != null ? int.tryParse(port) : null,
+  );
+  final String? finalCertPath = _getOverriddenValue<String>(
+    'TLS cert path',
+    fileConfig.https?.certPath,
+    tlsCertPath,
+  );
+  final String? finalCertKeyPath = _getOverriddenValue<String>(
+    'TLS cert key path',
+    fileConfig.https?.certKeyPath,
+    tlsCertKeyPath,
+  );
+  HttpsConfig? finalHttpsConfig;
+  if (finalCertPath != null || finalCertKeyPath != null || fileConfig.https != null) {
+    finalHttpsConfig = HttpsConfig(certPath: finalCertPath, certKeyPath: finalCertKeyPath);
+  } else {
+    finalHttpsConfig = null;
+  }
+  final finalHeaders = <String, String>{};
+  finalHeaders.addAll(fileConfig.headers);
+  if (headers != null && headers.isNotEmpty) {
+    for (final MapEntry<String, String> entry in headers.entries) {
+      if (fileConfig.headers.containsKey(entry.key)) {
+        globals.printStatus(
+          'Overriding headers "${entry.key}" from $devConfigFilePath ("${fileConfig.headers[entry.key]}") with command-line argument("${entry.value}").',
+        );
+      } else {
+        globals.printStatus('Adding header "${entry.key}" from command-line arguments.');
+      }
+    }
+    finalHeaders.addAll(headers);
+  }
+
+  return WebDevServerConfig(
+    host: finalHost,
+    port: finalPort,
+    https: finalHttpsConfig,
+    headers: finalHeaders,
+    proxy: fileConfig.proxy,
+  );
+}
+
+Future<int> resolvePort(int? port) async {
```

Comment:
The way this method works is a little bit weird, it either tries to find a free port or just checks that the incoming port is in range (but it doesn't check if the port is actually available). Do we even need the port < 0 > 65k validation? Do we even need this method? Is this just a fallback to `findFreePort`?